reason_4.0/lib/core/minisite_templates/modules/form_data/markup/default.php

```diff
@@ -0,0 +1,226 @@
+<?php
+/**
+ * @package reason
+ * @subpackage modules
+ */
+ 
+/**
+ * Register markup class
+ */
+$GLOBALS['form_data_markups']['minisite_templates/modules/form_data/markup/default.php'] = 'DefaultFormDataMarkup';
+
+/**
+ * A base class for form data display markup
+ *
+ * Usage:
+ * 1. Extend this class
+ * 2. Redefine get_markup() method to produce the desired markup
+ * 3. Specify the path to the the extended class file in the page type, like:
+ *
+ * 'my_form_display' => array(
+ * 		'main_post' => array(
+ * 			'module' => 'form_data',
+ * 			'markup' => 'path/to/markup/class.php',
+ * 		),
+ * ),
+ *
+ * 4. Associate a form with the page, or specify it in the page type for a firmer connection
+ *
+ * Example get_markup() definition for a form with 'Your Name' and 'Your Major' as labels.
+ *
+ * function get_markup()
+ * {
+ * 	$ret = '<ul>';
+ * 	foreach($this->get_form_data() as $row)
+ * 	{
+ * 		$ret .= '<li>'.$this->get_label_value('Your Name', $row).': '.$this->get_label_value('Your Major', $row).'</li>';
+ * 	}
+ * 	return $ret;
+ * }
+ */
+class DefaultFormDataMarkup
+{
+	protected $form;
+	protected $form_data;
+	protected $label_map;
+	
+	/**
+	 * Set the form entity
+	 *
+	 * @param object $form
+	 * @return void
+	 */
+	function set_form($form)
+	{
+		$this->form = $form;
+	}
+	/**
+	 * Get the form entity
+	 * @return object
+	 */
+	function get_form()
+	{
+		return $this->form;
+	}
+	
+	/**
+	 * Overload this method to interact with the head items object
+	 *
+	 * Examples:
+	 * $head_items->add_stylesheet('/path/to/css.css');
+	 * $head_items->add_javascript('/path/to/js.js');
+	 *
+	 * @param object $head_items
+	 * @return void
+	 */
+	function add_head_items($head_items)
+	{
```

Comment:
Should there be a default implementation of this method? It looks like this is used in the FormDataModule but this is basically a no-op.

Reply:
Yes, it's intentionally a no-op. The default version doensn't have any head ittems to add, so it does not add any head items. I would seem a little silly to have a stray empty css or js file just so this class could add a head item.

Reply:
An alternative would be to have a separate abstract class that defines this & other boilerplate methods, and extend it for the default markup.

Reply:
That's OK. My concern was as a developer I would likely extend this class (`DefaultFormDataMarkup `) for my own`FancyFormDataMarkup` class. I might want to add CSS and Javascript via
```
$head_items->add_stylesheet('/path/to/css.css');
$this->add_head_items($head_items);
```
I would expect that CSS to be added and not have to implement the `add_head_items()` method myself. Maybe the confusion for me is that the name `DefaultFormDataMarkup` and docs imply that this is a base class that should be extended.

I understand that this implementation (`DefaultFormDataMarkup`) has no head items. Maybe the default implementation of `add_head_items` can be no-op is the argument is null?

Reply:
I suspect that the semantics need to be tweaked.

The controller (e.g. the module) calls this method to pass a reference to the page's head items object (the intended argument) to the markup class, where it would (if it wanted to) add individual head items.

Here it is in use in local code (not part of this PR):

```php
function add_head_items($head_items)
	{
		$head_items->add_stylesheet('/global_stock/css/backstage/stars.less');
		$head_items->add_javascript('/global_stock/js/waypoints/lib/jquery.waypoints.min.js');
		$head_items->add_javascript('/global_stock/js/waypoints/lib/shortcuts/inview.min.js');
		$head_items->add_javascript('/global_stock/css/backstage/stars.js');
	}
```

Perhaps renaming this method would make the purpose clearer?

Not sure what to call it, though.

Reply:
I was not clear on how one might extend this base class. I don't want to get into reworking the plumbing. As long as this is discoverable for a developer, I think it's fine.

Reply:
I'll improve the documentation with a full example like the one I gave above.